### PR TITLE
Modify ReplayMemory

### PR DESCRIPTION
**Pull request by gorisanson**
Your Jupyter notebook was very helpful for me to learn about reinforcement learning and to train neural network myself. Thank you.

I open this pull request because I found something that could be improved in your code.

ReplayMemory is actually a ring buffer, but I found that it doesn't sample a state from leftmost indexes (indexes less than `self.agent_history_length - 1`) even if its elements count reaches it ring size. I modified somewhat of the codes so that ReplayMemory is able to sample a state from leftmost indexes once its elements count reaches its ring size.

And I also edited corresponding explanation "Finally ...... the array." without your permission. I think you can edit the explanation better because you are the author of this notebook:)

### CODE REVIEW

**Comment by fg91** [comment]
Hi Kyutae,

first of all, I want to apologize for taking a week to respond to your pr (I am currently in the final phase of writing my Ph.D. thesis so time is somewhat tight). I am very happy to hear that the tutorial was helpful to you!

Regarding your improvements:

- You are correct that the ring buffer in its current form does not sample from the first 4 frames. I am hesitant to merge your code at this point, however, because the method `_get_valid_indices` now contains lots of code duplication that is unelegant and will increase the time readers of this tutorial have to invest to understand what the code does. In my opinion, the (assumably small) performance benefit of having the full 1.000.000 frames instead of 999.996 useful frames in the replay memory does not outweigh this additional complexity to understand the tutorial.
- I am truly excited about your discovery that the rewards `gym` returns are not clipped! I remember looking at a few reward sequences of a random agent playing Breakout and not observing any rewards not equal to 0 or 1 so I assumed they were clipped and then later forgot about it. This makes sense though, probably the random agent I looked at only hit bricks in the first layer which truly returns a reward of 1 whereas bricks in later layers likely return much higher rewards. It's awesome, that you discovered this! I would be very interested in adding this to the code. Would you prepare a commit starting from my last commit that contains this change? Maybe put the function in its own cell below the Atari class (where the rewards are generated) and add the explanation:

Mnih et al. 2013 and Mnih et al. 2015 clipped the rewards explaining that "Since the scale of scores varies greatly from game to game, we fixed all positive rewards to be 1 and all negative rewards to be -1, leaving 0 rewards unchanged. Clipping the rewards in this manner limits the scale of the error derivatives and makes it easier to use the same learning rate across multiple games. At the same time, it could affect the performance of our agent since it cannot differentiate between rewards of different magnitude." Clipping the rewards makes the training converge faster and improves the performance in the example of Breakout.

Have you tried any other games? :)
Best regards,
Fabio

**Comment by gorisanson** [comment]
Hi Fabio,

I didn't know that my commits to the master branch of my fork repository reflected on this pull requests. Sorry about that. I'm still learning how to use git and github now. My intent for this pull request was just for ReplayMemory part. Thanks for considering reward clipping though.

Reading your kind explanation, now I agree with you about the code of ReplayMemory:)

I had thought that the rewards was already clipped too. But, after some training completed, I printed rewards on testing the agent just to check for that. And I realized that the rewards was not clipped. The rewards was different for breaking a brick of different layers like the following.

(scoreboard side)
fifth/sixth layer: 7 points per brick
third/fourth layer: 4 points per brick
first/second layer: 1 point per brick
(agent side)

I'll prepare a commit which include the reward clipping. And I think it is also needed to record evaluation score appropriately even if no evaluation game finished. Without this, there would be so many nan evaluation score because the training converges faster when the rewards were clipped. I'll include this to the commit too. 

I didn't tried other games from your code base. But I tried other two neural nets from your code base. The two other neural nets was these from Mnih et al. 2013 and Mnih et al. 2015. Actually, my first end goal was just to train these two neural nets and see the performance differences. But, after I found that reward clipping significantly improved these two neural nets, I wondered if reward clipping would also improve your duel network. So I tried your duel network too. The interesting fact I found during these tryings is that even Mnih et al. 2013 neural net (with reward clipping) achieves great agent performance for Breakout. This net converged slower than Mnih et al. 2015 neural net or your duel net (I used the same hyperparameters as your notebook throughout these tryings, with different hyperparametes Mnih et al. 2013 net maybe could converge faster?), but eventually achieved around 350 average training scores and around 400 evaluation scores (max evaluation score was 552).

Thanks for replying even if your time is so tight. Feel free to take time. These pr things have no due date, but your Ph. D. thesis certainly have due date:)

And thank you again for making this wonderful Jupyter notebook. Through these my own tryings I could feel that how much time and efforts would be taken to make this notebook. 

With thanks,
Kyutae

**Comment by gorisanson** [comment]
#3
